packages/authentication/src/provider.ts

```diff
@@ -0,0 +1,79 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as http from 'http';
+import {HttpErrors, inject, ParsedRequest} from '@loopback/core';
+import {Provider} from '@loopback/context';
+import {Strategy} from 'passport';
+import {StrategyAdapter} from './strategy-adapter';
+
+// interface definition of a function which accepts a request and returns an authenticated user
+export interface AuthenticateFn
+{
+  (request: ParsedRequest): Promise<UserProfile>;
+}
+
+// interface definition of a user profile http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+export interface UserProfile
+{
+  name?: string;
+  locale?: string;
+}
+
+/**
+ * @description Provider of a function which authenticates
+ * @example `context.bind('authentication_key').toProvider(AuthenticationProvider)`
+ */
+export class AuthenticationProvider implements Provider<AuthenticateFn> {
+  constructor(
+    @inject('authentication.strategy') private strategy: Strategy,
+  ) {}
+
+  /**
+   * @returns authenticateFn
+   */
+  value(): AuthenticateFn {
+    const authenticate: AuthenticateFn = async (request: ParsedRequest): Promise<UserProfile> => {
+      try {
+        return await getAuthenticatedUser(this.strategy, request);
+      } catch (err) {
+        return Promise.reject(err);
```

Comment:
This is not needed. The compiler/runtime will automatically catch errors thrown by async functions and convert them to promise rejections.

```
$ node
> (async function foo() { throw new Error('oops'); })()
Promise {
  <rejected> Error: oops
    at foo (repl:1:31)
    at repl:1:52
   (...)
```

Here is my proposal:

```ts
value(): AuthenticateFn {
  return async authenticateRequest(request: ParsedRequest): Promise<UserProfile>  => {
    return await getAuthenticatedUser(this.strategy, request);
  }
}
```

BTW I think the type specification of the return type (`:Promise<UserProfile>`) is redundant, TS compiler should be able to infer it automatically for us.